dbms/src/Common/ThreadProfileEvents.cpp

```diff
@@ -0,0 +1,218 @@
+#include <Common/ThreadProfileEvents.h>
+
+#if defined(__linux__)
+#include <unistd.h>
+#include <linux/perf_event.h>
+#include <syscall.h>
+#include <sys/ioctl.h>
+#include <cerrno>
+#endif
+
+namespace DB {
```

Comment:
It may be nice to have separate methods (Start, Stop, Reset) to support the `ioctl` options on a profilers like `PERF_EVENT_IOC_RESET`, `PERF_EVENT_IOC_ENABLE` and `PERF_EVENT_IOC_DISABLE`. They sort of exist in the current implementation, but it would be nice if you could have those extra controls.

Reply:
Can you please describe how those (Start, Stop, Reset) methods can be called from outside of the ClickHouse process (or otherwise, I don't understand when they are useful within ClickHouse codebase)? Probably, due to the lack of knowledge and/or experience I do not understand the phrase "to support the `ioctl` options on a profilers like `PERF_EVENT_IOC_RESET`, `PERF_EVENT_IOC_ENABLE` and `PERF_EVENT_IOC_DISABLE`" and googling didn't help with that, unfortunately. So, will be glad to hear about examples (or just the same thing but with a bit of explanation) when these methods can be useful

Reply:
Ideally they would be used when a user updates ClickHouse settings. For example, all the `PERF_COUNT_HW_*` events could have a setting of `perf_hardware_events_enabled` or something similar that would also expose all these metrics via graphite or a system table. Perf counters are increasing while they are enabled (however they can wrap uint64s). The difficulty becomes when you want to restart profiling, if you disable the configuration should the counter be reset to zero or continue from the last value (this is what the ioctls would handle). Then you could have another configuration field such as `perf_hardware_events_reset_on_enable` to control whether an profiler is reset when it is enabled.